src/chunk.c

```diff
@@ -3292,111 +3274,48 @@ ts_chunk_create_fks(const Hypertable *ht, const Chunk *const chunk)
 	}
 }
 
-static ScanTupleResult
-chunk_tuple_update_schema_and_table(TupleInfo *ti, void *data)
+/* Chunk catalog updates are done in three steps.
+ * This is achieved by following this sequence:
+ *   1: call lock_chunk_tuple and which locks the tuple and fills TID and data
+ *   2: make changes to the data
+ *   3: call chunk_update_catalog_tuple with the TID and updated data
+ *
+ * This is equivalent to SELECT for UPDATE, followed by UPDATE
+ *
+ * All callers who want to update chunk tuples should respect this so that locks
+ * are acquired correctly.
+ *
+ */
+static void
+chunk_update_catalog_tuple(ItemPointer tid, FormData_chunk *update)
 {
-	FormData_chunk form;
-	FormData_chunk *update = data;
-	CatalogSecurityContext sec_ctx;
 	HeapTuple new_tuple;
-
-	ts_chunk_formdata_fill(&form, ti);
-
-	namestrcpy(&form.schema_name, NameStr(update->schema_name));
-	namestrcpy(&form.table_name, NameStr(update->table_name));
-
-	new_tuple = chunk_formdata_make_tuple(&form, ts_scanner_get_tupledesc(ti));
-
-	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
-	ts_catalog_update_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti), new_tuple);
-	ts_catalog_restore_user(&sec_ctx);
-	heap_freetuple(new_tuple);
-	return SCAN_DONE;
-}
-
-static ScanTupleResult
-chunk_tuple_update_status(TupleInfo *ti, void *data)
-{
-	FormData_chunk form;
-	FormData_chunk *update = data;
 	CatalogSecurityContext sec_ctx;
-	HeapTuple new_tuple;
+	Catalog *catalog = ts_catalog_get();
+	Oid table = catalog_get_table_id(catalog, CHUNK);
+	Relation chunk_rel = relation_open(table, RowExclusiveLock);
 
-	ts_chunk_formdata_fill(&form, ti);
-	form.status = update->status;
-	new_tuple = chunk_formdata_make_tuple(&form, ts_scanner_get_tupledesc(ti));
+	new_tuple = chunk_formdata_make_tuple(update, chunk_rel->rd_att);
 
 	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
-	ts_catalog_update_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti), new_tuple);
+	ts_catalog_update_tid(chunk_rel, tid, new_tuple);
 	ts_catalog_restore_user(&sec_ctx);
 	heap_freetuple(new_tuple);
-	return SCAN_DONE;
+	relation_close(chunk_rel, NoLock);
 }
 
 static bool
```

Comment:
Could you please add a comment to lock_chunk_tuple?

this function locks the  timescaledb_catalog.chunk tuple (corresponding to chunk_id ) in LockTupleExclusiveMode. It blocks till the lock is acquired.
The tid and data (corresponding to the locked tuple) are returned via tid and form  arguments.
Anyone updating/deleting a chunk entry from the catalog table is expected to first call this function.
Refer to chunk_update_catalog_tuple() for more details,